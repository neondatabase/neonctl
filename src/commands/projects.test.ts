import { afterAll, describe, expect, test } from 'vitest';
import { readFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { testCliCommand } from '../test_utils/test_cli_command.js';

const CONTEXT_FILE = join(tmpdir(), `neon_project_create_ctx_${Date.now()}`);

describe('projects', () => {
  testCliCommand({
    name: 'list',
    args: ['projects', 'list'],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'create',
    args: ['projects', 'create', '--name', 'test_project'],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'create with database and role',
    args: [
      'projects',
      'create',
      '--name',
      'test_project',
      '--database',
      'test_db',
      '--role',
      'test_role',
    ],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'create and connect with psql',
    args: ['projects', 'create', '--name', 'test_project', '--psql'],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'create and connect with psql and psql args',
    args: [
      'projects',
      'create',
      '--name',
      'test_project',
      '--psql',
      '--',
      '-c',
      'SELECT 1',
    ],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'create project with setting the context',
    args: [
      'projects',
      'create',
      '--name',
      'test_project',
      '--context-file',
      CONTEXT_FILE,
      '--set-context',
    ],
    expected: {
      snapshot: true,
    },
  });

  afterAll(() => {
    rmSync(CONTEXT_FILE);
  });

  test('context file should exist and contain the project id', () => {
    expect(readFileSync(CONTEXT_FILE, 'utf-8')).toContain('new-project-123456');
  });

  testCliCommand({
    name: 'delete',
    args: ['projects', 'delete', 'test'],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'update name',
    args: ['projects', 'update', 'test', '--name', 'test_project_new_name'],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'update ip allow',
    args: [
      'projects',
      'update',
      'test',
      '--ip-allow',
      '127.0.0.1',
      '192.168.1.2/22',
      '--ip-primary-only',
    ],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'update ip allow primary only flag',
    args: ['projects', 'update', 'test', '--ip-primary-only', 'false'],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'update ip allow remove',
    args: ['projects', 'update', 'test', '--ip-allow'],
    expected: {
      snapshot: true,
    },
  });

  testCliCommand({
    name: 'get',
    args: ['projects', 'get', 'test'],
    expected: {
      snapshot: true,
    },
  });
});
