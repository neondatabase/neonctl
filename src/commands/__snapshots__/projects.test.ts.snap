// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`projects > create 1`] = `
"project:
  id: new-project-123456
  name: test_project
  created_at: 2021-01-01T00:00:00.000Z
connection_uris:
  - connection_uri: postgres://localhost:5432/test_project
"
`;

exports[`projects > create and connect with psql 1`] = `
"project:
  id: new-project-123456
  name: test_project
  created_at: 2021-01-01T00:00:00.000Z
connection_uris:
  - connection_uri: postgres://localhost:5432/test_project
{"psql-cli-args":["postgres://localhost:5432/test_project"]}"
`;

exports[`projects > create and connect with psql and psql args 1`] = `
"project:
  id: new-project-123456
  name: test_project
  created_at: 2021-01-01T00:00:00.000Z
connection_uris:
  - connection_uri: postgres://localhost:5432/test_project
{"psql-cli-args":["postgres://localhost:5432/test_project","-c","SELECT 1"]}"
`;

exports[`projects > create project with default autoscaled CU 1`] = `
"project:
  id: new-project-123456
  name: test_project
  created_at: 2021-01-01T00:00:00.000Z
connection_uris:
  - connection_uri: postgres://localhost:5432/test_project
"
`;

exports[`projects > create project with default fixed size CU 1`] = `
"project:
  id: new-project-123456
  name: test_project
  created_at: 2021-01-01T00:00:00.000Z
connection_uris:
  - connection_uri: postgres://localhost:5432/test_project
"
`;

exports[`projects > create project with setting the context 1`] = `
"project:
  id: new-project-123456
  name: test_project
  created_at: 2021-01-01T00:00:00.000Z
connection_uris:
  - connection_uri: postgres://localhost:5432/test_project
"
`;

exports[`projects > create with database and role 1`] = `
"project:
  id: new-project-123456
  name: test_project
  created_at: 2021-01-01T00:00:00.000Z
connection_uris:
  - connection_uri: postgres://localhost:5432/test_project
"
`;

exports[`projects > create with org id 1`] = `
"project:
  id: new-project-789012
  name: test_project
  created_at: 2022-01-01T00:00:00.000Z
  org_id: org-2
connection_uris:
  - connection_uri: postgres://localhost:5432/test_project
"
`;

exports[`projects > delete 1`] = `
"id: test
name: test
create_at: 2018-01-01T00:00:00.000Z
"
`;

exports[`projects > get 1`] = `
"id: test
name: test_project
created_at: 2019-01-01T00:00:00Z
settings:
  allowed_ips:
    ips:
      - 192.168.1.1
    primary_branch_only: false
"
`;

exports[`projects > list 1`] = `
"projects:
  - id: 1
    name: Project_1
    created_at: 2019-01-01T00:00:00.000Z
    updated_at: 2019-01-01T00:00:00.000Z
  - id: 2
    name: Project_2
    created_at: 2019-01-01T00:00:00.000Z
    updated_at: 2019-01-01T00:00:00.000Z
  - id: 3
    name: Project_3
    created_at: 2019-01-01T00:00:00.000Z
    updated_at: 2019-01-01T00:00:00.000Z
shared_with_me:
  - id: adj-noun-12401747
    region_id: aws-us-east-2
    name: Shared Project
    created_at: 2024-04-03T04:45:46Z
    updated_at: 2024-04-11T16:13:43Z
"
`;

exports[`projects > list with org id 1`] = `
"- id: 4
  name: Project_4
  created_at: 2019-01-01T00:00:00.000Z
  updated_at: 2019-01-01T00:00:00.000Z
  org_id: org-2
- id: 5
  name: Project_5
  created_at: 2019-01-01T00:00:00.000Z
  updated_at: 2019-01-01T00:00:00.000Z
  org_id: org-2
- id: 6
  name: Project_6
  created_at: 2019-01-01T00:00:00.000Z
  updated_at: 2019-01-01T00:00:00.000Z
  org_id: org-2
"
`;

exports[`projects > update ip allow 1`] = `
"id: test
name: test_project
created_at: 2019-01-01T00:00:00Z
settings:
  allowed_ips:
    ips:
      - 127.0.0.1
      - 192.168.1.2/22
    primary_branch_only: true
"
`;

exports[`projects > update ip allow primary only flag 1`] = `
"id: test
name: test_project
created_at: 2019-01-01T00:00:00Z
settings:
  allowed_ips:
    ips:
      - 192.168.1.1
    primary_branch_only: false
"
`;

exports[`projects > update ip allow remove 1`] = `
"id: test
name: test_project
created_at: 2019-01-01T00:00:00Z
settings:
  allowed_ips:
    ips: []
    primary_branch_only: false
"
`;

exports[`projects > update name 1`] = `
"id: test
name: test_project_new_name
created_at: 2019-01-01T00:00:00Z
settings:
  allowed_ips:
    ips:
      - 192.168.1.1
    primary_branch_only: false
"
`;

exports[`projects > update project with default autoscaled CU 1`] = `
"id: test
name: test_project_with_autoscaling
created_at: 2019-01-01T00:00:00Z
settings:
  allowed_ips:
    ips:
      - 192.168.1.1
    primary_branch_only: false
"
`;

exports[`projects > update project with default fixed size CU 1`] = `
"id: test
name: test_project_with_fixed_cu
created_at: 2019-01-01T00:00:00Z
settings:
  allowed_ips:
    ips:
      - 192.168.1.1
    primary_branch_only: false
"
`;
